Add CLI data integrity repair tool

anvit opened the pull request:
Add a CLI tool for performing data integrity checks and repairing or deleting broken records. 

(Adapted from @jraddaoui's data integrity script into a CLI task)

anvit committed:
Add CLI data integrity repair tool

Add a CLI tool for performing data integrity checks and repairing or
deleting broken records.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+                ++$fixed;
+            }
+
+            $this->log(sprintf("  - %s: %d\n", $class, $fixed));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+        $updated = QubitPdo::modify($sql);
+        $this->logSection('data-integrity-repair', sprintf("Updating terms without parent id: %d\n", $updated));
+
+        $this->log("Checking descriptions integrity:\n");
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          LEFT JOIN object o ON io.id=o.id
+          WHERE io.id<>1
+          AND o.id IS NULL;';
+        $this->log(sprintf("  - Descriptions without object row: %d\n", QubitPdo::fetchColumn($sql)));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          FROM information_object
+          WHERE id<>1
+          AND parent_id IS NULL;';
+        $this->log(sprintf("  - Descriptions without parent id: %d\n", QubitPdo::fetchColumn($sql)));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          LEFT JOIN information_object p ON io.parent_id=p.id
+          WHERE io.id<>1
+          AND p.id IS NULL;';
+        $this->log(sprintf("  - Descriptions without parent: %d\n", QubitPdo::fetchColumn($sql)));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          LEFT JOIN status st ON io.id=st.object_id AND st.type_id=158
+          WHERE io.id<>1
+          AND st.status_id IS NULL;';
+        $this->log(sprintf("  - Descriptions without publication status: %d\n", QubitPdo::fetchColumn($sql)));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+        $affectedIos = QubitPdo::fetchAll(
+        $sql, [], ['fetchMode' => PDO::FETCH_ASSOC]
+        );
+        $this->log(sprintf("  - Affected descriptions: %d\n", count($affectedIos)));
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+        $this->log(sprintf("  - Affected descriptions: %d\n", count($affectedIos)));
+
+        if (0 == count($affectedIos)) {
+            $this->log("All descriptions seem to be okay.\n");
```
Use `$this->logSection()`.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          OR p.id IS NULL
+          OR st.id IS NULL
+          OR st.status_id IS NULL);';
+        $affectedIos = QubitPdo::fetchAll(
+        $sql, [], ['fetchMode' => PDO::FETCH_ASSOC]
+        );
```
Indentation is wrong, it would fit in one line too.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+            $this->logSection('data-integrity-repair', sprintf("  - Affected descriptions (including descendants): %d\n", count($affectedIosAndDescendantIds)));
+
+            switch ($options['fix']) {
+        case 'fix':
+          $this->report($filename, $affectedIosById, $affectedIosAndDescendantIds);
+          $this->fix($affectedIosById);
+
+          break;
+
+        case 'delete':
+          $this->report($filename, $affectedIosById, $affectedIosAndDescendantIds);
+          $this->deleteDescriptions($affectedIosById, $affectedIosAndDescendantIds);
+
+          break;
+
+        default:
+          $this->report($filename, $affectedIosById, $affectedIosAndDescendantIds);
+
+          break;
+      }
```
Indentation is wrong here too, how is this not found by the syntax checker?

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+        $this->briefDescription = 'Attempt data integrity repair';
+        $this->detailedDescription = <<<'EOF'
+Attempt to repair data integrity. It does the following:
+- Add missing object rows for all resources extending QubitObject
```
Use `Adds` for consistency.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+- Add missing object rows for all resources extending QubitObject
+- Regenerates slugs to use them in CSV report
+- Adds missing parent ids to terms
+- Checks descriptions with missing data and provides options for attempting to generate a list, fix them, or delete them
```
Include `Re-builds the nested sets` or something like that in this list.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+
+            switch ($options['fix']) {
+        case 'fix':
+          $this->report($filename, $affectedIosById, $affectedIosAndDescendantIds);
```
Report happens in all cases, you could remove the breaks and make all cases fall down to the default, not needing to call the report generation on each case. You could also move the report generation before the switch and remove the default case.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+          (id, class_name, created_at, updated_at, serial_number)
+          VALUES
+          (:id, :class, now(), now(), 0);';
+        QubitPdo::modify(
+        $sql, [':id' => $id, ':class' => $class]
+        );
```
Indentation/one-line.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+
+        // Find children
+        $sql = 'SELECT id FROM information_object WHERE parent_id=:id;';
+        $children = QubitPdo::fetchAll(
+        $sql, [':id' => $id], ['fetchMode' => PDO::FETCH_COLUMN]
+        );
```
Indentation/one-line.

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+    protected function configure()
+    {
+        $this->addArguments([
+            new sfCommandArgument('filename', sfCommandArgument::OPTIONAL, 'A name for the generated CSV report file', 'affected-records.csv'),
```
A filepath instead of a name and clarify it requires the `.csv` extension (so we could remove the check below).

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,359 @@
+        $affectedIosAndDescendantIds[] = $id;
+    }
+
+    private function stringEndsWith($string, $pattern)
+    {
+        return 0 === substr_compare($string, $pattern, -strlen($pattern));
+    }
+
+    private function report($filename, $affectedIosById, $affectedIosAndDescendantIds)
+    {
+        if (!($this->stringEndsWith($filename, '.csv') || $this->stringEndsWith($filename, '.CSV'))) {
+            $filename = sprintf('%s.csv', $filename);
+        }
```
Maybe clarifying a CSV path is required, this is no lo longer needed.

jraddaoui reviewed (changes requested):
Thank you for adapting this @anvit and sorry for the delay on the review.

jraddaoui commented:
I don't know what are the plans for documentation, but we should explain somewhere the meaning of the CSV and the issues that can be found there.

anvit committed:
Fix indentation and clean up code

- Fix indentation
- Change calls to log to logSection
- Move duplicated calls to report outside the switch
- Remove csv filename extension check
- Update help message, and clarify the .csv file extension

anvit commented:
@jraddaoui Just addressed the feedback! Not sure why PHP-CS fixer didn't catch indentation issues because it seems to be catching other things in the file just fine, but those should be fixed now

jraddaoui commented on lib/task/tools/dataIntegrityRepairTask.class.php:
```diff
@@ -0,0 +1,345 @@
+        $affectedIosAndDescendantIds[] = $id;
+    }
+
+    private function stringEndsWith($string, $pattern)
+    {
+        return 0 === substr_compare($string, $pattern, -strlen($pattern));
+    }
```
I think you can remove this function now.

jraddaoui reviewed (approved):
Thanks @anvit!

anvit committed:
Remove unused function

Remove the stringEndsWith function which was unused

melaniekung reviewed (approved):
Looks good to me! Thanks Anvit!
